framework/calico_framework.py

```diff
@@ -0,0 +1,666 @@
+                update.task_id.value)
+            driver.abort()
+
+        # Update the calico_task with info from TaskInfo
+        calico_task.state = update.state
+        for network_info in update.container_status.network_infos:
+            if calico_task.ip:
+                if network_info.ip_address != calico_task.ip:
+                    self.kill_test(calico_task.test,
+                                   "Assigned IP (%s) doesn't match "
+                                   "desired IP of %s" % \
+                    (network_info.ip_address, calico_task))
+            else:
+                calico_task.ip = network_info.ip_address
+
+        _log.info("TASK_UPDATE - %s: %s", \
+                  mesos_pb2.TaskState.Name(calico_task.state),
+                  calico_task)
+
+        if not update.healthy:
+            _log.error(
+                "\t%s is in unexpected state %s with message '%s'",
+                calico_task,
+                mesos_pb2.TaskState.Name(update.state),
+                update.message)
+            _log.error("\tData:  %s", repr(str(update.data)))
+            _log.error("\tSent by: %s",
+                   mesos_pb2.TaskStatus.Source.Name(update.source))
+            _log.error("\tReason: %s",
+                   mesos_pb2.TaskStatus.Reason.Name(update.reason))
+            _log.error("\tMessage: %s", update.message)
+            _log.error("\tHealthy: %s", update.healthy)
+
+            self.kill_test(calico_task.test, update.message)
+            return
+
+
+        # Check for good update
+        if update.state == mesos_pb2.TASK_FINISHED:
+            # If its a sleep task, check that its pingers have finished as well
+            if type(calico_task) == SleepTask:
+                calico_task_targeters = [task for task in calico_task.test.tasks
+                                         if type(task) == PingTask and
+                                         calico_task in (
+                                         task.can_ping_targets + task.cant_ping_targets)]
+
+                for calico_task_targeter in calico_task_targeters:
+                    if calico_task_targeter.state is not mesos_pb2.TASK_FINISHED:
+                        self.kill_test(calico_task.test,
+                                       "A Sleep task finished before "
+                                       "its Pinger did.")
+                        return
+
+        # Check for test completion, or reset timeout
+        for task in calico_task.test.tasks:
+            if task.state != mesos_pb2.TASK_FINISHED:
+                calico_task.test.restart_timeout()
+                break
+        else:
+            _log.info("\tTEST_COMPLETE: %s", calico_task.test)
+            calico_task.test.state = TestState.Complete
+            calico_task.test.timeout = None
+            # Need to remove the entry in test_by_slave_id
+            slaves_to_unreserve = []
```

Comment:
I assume you're using this repeated loop since you want to modify `test_by_slave_id`.  A more straightforward way to do this is to make a temporary copy and iterate over that.  E.g.

```
for slave_id, test in list(self.test_by_slave_id.iteritems()):
    if test == calico_task.test:
        del self.test_by_slave_id[slave_id]
```